### ISSUE TITLE
Remove Mongo from Phalcon

### ISSUE THREAD

**Issue by niden** [phalcon/cphalcon#13697]
Since there is no support for the existing driver for PHP, we have to remove Mongo from this release.

There is a plan to re-introduce it in the future as a storage adapter and with that we can re-introduce the `Mvc\Collection` class or something similar

**Comment by ViltusVilks** [comment]
What do You mean by removing Mongo? Remove direct "usage" of mongo in .zep and also remove ODM completely in 4.0 release?

**Comment by niden** [comment]
@ViltusVilks Remove the current driver (Mongo driver) that Phalcon uses to interact with Mongo because it has been deprecated. That will mean that the ODM will not work for a period of time until the new driver is introduced.

If we can do this in one go that would be great.

**Comment by ViltusVilks** [comment]
Huh, deprecated? What is this? https://pecl.php.net/package/mongodb
Or I do not understand what is deprecated (Phalcon support for ODM/Mongo compatibility, or Mongo decided to drop support for MongoDB native driver (NETLIB)?)

I am working on dokerized Phalcon-ODM-only project to make slim, fast and scalable image. Without ODM project is dead :(

**Comment by niden** [comment]
@ViltusVilks It is the driver itself. Phalcon has been using this driver:

http://php.net/manual/en/book.mongo.php

and we need to change it to this driver

http://php.net/manual/en/set.mongodb.php

**Comment by niden** [comment]
It is a matter of time and resources really. We have not "removed" anything with regards to the Mongo functionality. If you have the first driver installed then it will work I think. I really don't know the changes that 7.2 brought in that aspect of PHP.

Removing mongo functionality including the ODM in one release and reintroducing it properly tested and upgraded is the plan. It has not materialized yet since we have not finished the release - still in alpha

**Comment by ViltusVilks** [comment]
Ah... this is old one driver! 
Just for information - I am using **[incubator](https://github.com/phalcon/incubator)** github project, which overrides collection and uses new driver. I also utilize GridFS features. 

And everything works fine php7.2.13 + 3.4.2 on Alpine 3.8 :)
Have You checked that?

Shortcut - base collection class which is replacing Mvc\Collection. (i am using in Models namespace)
https://github.com/phalcon/incubator/blob/master/Library/Phalcon/Mvc/MongoCollection.php

If that stuff can be "ported" to .zep ... on re-implemented /cleaned.

**Comment by wajdijurry** [comment]
I am using Phalcon MongoCollection and Client component to connect to
mongo. It works as a charm.
Phalcon version 3.4.1

**Comment by niden** [comment]
@ViltusVilks Yes that is the plan. To upgrade the driver and clean things up. Right now we don't have a lot of time so this is why I wrote we need to remove it and bring it back fresh.

The "removal" will be temporary of course. We are not killing the component.

**Comment by npfedwards** [comment]
@niden I thought we'd always had to use the incubator version and that was using the new driver (since php7.0?). I'm a complete mongo convert so this is kinda important to me. Very willing to take a look at this once I've got used to writing the tests etc.

**Comment by ViltusVilks** [comment]
Yep, incubator is good and can be taken as an "example".

FYI: Just tryied to convert simple ODM models/code for Phalcon4 by modifing incubator code, it does not work, because of strict-type on interfaces on return types :( 

Basically Phalcon4 should provide generic collection+manager with possibility to attach any backend adapter capable to handle Document store and query (incl. Solr, Elastic ...)

**Comment by npfedwards** [comment]
@niden Sure. I'll have a go making sure the current incubator code is tested (I've been using it for a while so I'm pretty sure it works ;)). And then given I want to expand my knowledge I'll have a go porting it to Zephir. I expect it to be a few week project for me but expect a pull request in a while.

**Comment by ViltusVilks** [comment]
Some news.
I have got working Phalcon v4.alpha2 with Incubator adapter and some small changes

Of cause you need do some code fixes like getSource() : string and etc typed returns for interface functions, but quering and save works fine :)

**CollectionInterface**, propper return type
```
/**
	 * Allows to query the first record that match the specified conditions
	 */
	public static function findFirst(array parameters = null) -> <CollectionInterface> | null;
```

**MongoCollecto@Incubator**, do the same
```
    // @codingStandardsIgnoreStart
    static protected $_disableEvents = false;
    // @codingStandardsIgnoreEnd

    public static function findFirst(array $parameters = null) : \Phalcon\Mvc\CollectionInterface
...
```

**Comment by v00v** [comment]
@niden This is similar to what I was requesting as NFR [1] Its a great idea having ``Phalcon\Collection`` with CRUD support. It can use similar functions like in ODM [2] I specificly requested JSON adapter (flat-file-storage) because this is my own preference. But it can easily be extended to YAML and pure flat-file-storage similar to cache in Phalcon now but with much more control over it like sub-folders and search.

[1] https://github.com/phalcon/cphalcon/issues/13783
[2] https://docs.phalconphp.com/3.2/en/db-odm

**Comment by SidRoberts** [comment]
Does anyone have any updates on this?

**Comment by niden** [comment]
@SidRoberts From what I mentioned in a comment above, the storage adapters and the structure is there in the PSR-16 branch (my fork). Once that gets merged into 4.0.x we will have the structure to implement a mongo adapter using the latest driver. After that we can implement the Collection component that is aligned with the Resultset one.

There were a couple of members of the community that have been working on setting up the mongo collection with the latest mongo driver offering better compatibility with the latest and greatest mongo functionality.

Honestly I don't know about the timing on this i.e. whether we will have time to replace the mongo driver. For v4 we just might remove it and introduce it to v4.1

**Comment by tacxticx88** [comment]
Hello, I propose you a (temporary) implementation while waiting for the total rewriting of the collections with MongoDb / PHP7. There is probably still work to be done, and unit tests must be done. The goal is to offer support for the new MongoDb driver for corporate projects
 - https://github.com/toroia/phalconator/tree/develop
